### ISSUE TITLE
Custom tags with hyphen in *.jsx files crash JavaScript analysis

### ISSUE THREAD

**Issue by catalan-adobe** [SonarSource/SonarJS#1031]
**I want to report a bug.**

**SonarJS version:** 3.3
**SonarQube version:** 6.3

**Reproducer** 
```
render() {
         return <g>
             <if condition={ this.src }>
                 <img { ...this.undeclaredAttributes() }
                     class= { this._thumbnailClasses }
                     src = { this.src }
                     alt="" />
             </if>
             <else>
                 <my-icon
                    ^
                     icon={ this.icon }
                     { ...this.undeclaredAttributes() }
                     class={ this._thumbnailClasses}
                     alt="" />
             </else>
         </g>;
     }
```

**Logs**
```
12:10:06 [ERROR] Unable to parse file: [...]/components/cell/CellThumbnail.jsx
12:10:06 [ERROR] Parse error at line 185 column 23:
12:10:06 
12:10:06 175: 
12:10:06 176:     render() {
12:10:06 177:         return <g>
12:10:06 178:             <if condition={ this.src }>
12:10:06 179:                 <img { ...this.undeclaredAttributes() }
12:10:06 180:                     class= { this._thumbnailClasses }
12:10:06 181:                     src = { this.src }
12:10:06 182:                     alt="" />
12:10:06 183:             </if>
12:10:06 184:             <else>
12:10:06 185:                 <my-icon
12:10:06                         ^
12:10:06 186:                     icon={ this.icon }
12:10:06 187:                     { ...this.undeclaredAttributes() }
12:10:06 188:                     class={ this._thumbnailClasses}
12:10:06 189:                     alt="" />
12:10:06 190:             </else>
12:10:06 191:         </g>;
12:10:06 192:     }
12:10:06 193: }
12:10:06 194: 
```

**Expected behavior**

File should be parsed without issue

**Comment by legalcodes** [comment]
+1 also experiencing this issue for any file using a hyphen in a jsx tag. 

```
106:               <foo-bar
                              ^
```

**Comment by vilchik-elena** [comment]
Hi,

Thanks for reporting this issue. Indeed it looks like our JSX parser allows only letters in element name.

Currently we are gradually migrating SonarJS from our own parser to espree and babel parsers, so it's unlikely we will fix this problem in our parser.

Update to latest version of SonarJS (5.0). While you will continue seeing this parsing error, babel parser should parse it correctly and some part of our rules will be applied already (but not metrics and highlighting yet :( ). With next releases we will migrate more and more functionality.

I keep issue open, if we see that many users are affected, we might consider patching our parser.

**Comment by holblin** [comment]
Hi, today I had the same issue. Sadly we used some framework that is not under our control and have theses kind of tags.
What is the status of the issue? Id it possible to allow `-` and `:` in the tagName?

**Comment by guillaume-brialon** [comment]
Hi, I add my stone to the wall. I use third party web components having a `-` in their name. Is there a workaround or a config trick to overcome this error?

**Comment by holblin** [comment]
Could someone try this fix:
https://github.com/SonarSource/SonarJS/blob/3af4da1a1aca01274edb375ab69b0eb3e1a4ad58/javascript-frontend/src/main/java/org/sonar/javascript/parser/EcmaScriptLexer.java#L447

Replace:
>     b.rule(JSX_HTML_TAG).is(SPACING, b.regexp("^(?!this)[a-z][\\w]*"));
By:
>     b.rule(JSX_HTML_TAG).is(SPACING, b.regexp("^(?!this)[a-z][\\w-:]*"));

**Comment by holblin** [comment]
For us and our custom JSX it's a little bit more complicated since we could also have `:` inside JSX attributes. 🤦‍♂ 

So I also changed: `javascript-frontend/src/main/java/org/sonar/javascript/lexer/JavaScriptLexer.java`

I replace:
```private static final String JSX_IDENTIFIER_PART = "(?:" + IDENTIFIER_START + "|[" + UNICODE_COMBINING_MARK + UNICODE_DIGIT + UNICODE_CONNECTOR_PUNCTUATION + "|-])";```
By:
```private static final String JSX_IDENTIFIER_PART = "(?:" + IDENTIFIER_START + "|[" + UNICODE_COMBINING_MARK + UNICODE_DIGIT + UNICODE_CONNECTOR_PUNCTUATION + "|-]|:)";```
And that works 🎉 

But now I still have an other issue 😢 . Our specific JSX has this kind of tags which fails:
```
180:           <repeat collection={ FLAGS } as={ flag, name }>
                                                     ^
```

I don't have any idea of how to fix that. I will need some help so if you have any idea... 🤷‍♂

**Comment by vilchik-elena** [comment]
@holblin is such custom JSX parsed by babel?

**Comment by holblin** [comment]
yes, with a specific babel plugin. I need to grab more details on this.

**Comment by holblin** [comment]
Could we reopen it?
The fix in #1242 is partial, I know that I have been a little bit busy but I should be able to provide a complete fix during the next week. But I didn't have test associated even if we run it for few weeks now without problems.

**Comment by andrea-guarino-sonarsource** [comment]
Hi @holblin ,
I'd suggest to open a new ticket with a small reproducer of the issue.
And, of course, you're welcome to contribute with a Pull Request 🙂
